packages/contracts/contracts/GenArt721CoreV4.sol

```diff
@@ -0,0 +1,2010 @@
+
+    struct Project {
+        uint24 invocations;
+        uint24 maxInvocations;
+        uint24 scriptCount;
+        // max uint64 ~= 1.8e19 sec ~= 570 billion years
+        uint64 completedTimestamp;
+        bool active;
+        bool paused;
+        string name;
+        string artist;
+        address descriptionAddress;
+        string website;
+        string license;
+        string projectBaseURI;
+        bytes32 scriptTypeAndVersion;
+        string aspectRatio;
+        // mapping from script index to address storing script in bytecode
+        mapping(uint256 => address) scriptBytecodeAddresses;
+        address artistContractAddress;
+    }
+
+    mapping(uint256 => Project) projects;
+
+    /// packed struct containing project financial information
+    struct ProjectFinance {
+        address payable additionalPayeePrimarySales;
+        // packed uint: max of 95, max uint8 = 255
+        uint8 secondaryMarketRoyaltyPercentage;
+        address payable additionalPayeeSecondarySales;
+        // packed uint: max of 100, max uint8 = 255
+        uint8 additionalPayeeSecondarySalesPercentage;
+        address payable artistAddress;
+        // packed uint: max of 100, max uint8 = 255
+        uint8 additionalPayeePrimarySalesPercentage;
+    }
+    // Project financials mapping
+    mapping(uint256 => ProjectFinance) projectIdToFinancials;
+
+    /// hash of artist's proposed payment updates to be approved by admin
+    mapping(uint256 => bytes32) public proposedArtistAddressesAndSplitsHash;
+
+    /// Art Blocks payment address for all primary sales revenues (packed)
+    address payable public artblocksPrimarySalesAddress;
+    /// Percentage of primary sales revenue allocated to Art Blocks (packed)
+    // packed uint: max of 25, max uint8 = 255
+    uint8 private _artblocksPrimarySalesPercentage = 10;
+
+    /// Art Blocks payment address for all secondary sales royalty revenues
+    address payable public artblocksSecondarySalesAddress;
+    /// Basis Points of secondary sales royalties allocated to Art Blocks
+    uint256 public artblocksSecondarySalesBPS = 250;
+
+    /// single minter allowed for this core contract
+    address public minterContract;
+
+    /// starting (initial) project ID on this contract
+    uint256 public immutable startingProjectId;
+
+    /// next project ID to be created
+    uint248 private _nextProjectId;
+
+    /// bool indicating if adding new projects is forbidden;
+    /// default behavior is to allow new projects
+    bool public newProjectsForbidden;
+
+    /// version & type of this core contract
+    string public constant coreVersion = "v4.0.0";
+    string public constant coreType = "GenArt721CoreV4";
+
+    /// default base URI to initialize all new project projectBaseURI values to
+    string public defaultBaseURI;
+
+    // Maps tokenIds to projectIds because cannot just divide by 1M anymore
+    mapping(uint => uint) public tokenIdToProjectIds;
```

Comment:
This mapping is to ensure we keep track of tokens in the case of non-serial minting. Here is the scenario - 

1. Create drop on AB, say for 100 tokens.
2. People mint 3 tokens on AB.
3. Creator mints a token normally on the same contract
4. People mint 2 more tokens on AB.

Then this AB contract only manages tokens 1-3 and 4-5. This is why we must keep track.